keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+
+There can be additional resource allocation conditions and policies handled
+externally to the kubelet, e.g.,
+
+* Alternative resource allocator implemented by the CRI which needs to reject
+  pods in conditions similar to the built-in resource managers.
+
+These can also benefit from a pluggable admission handler to reject pods before
+they are admitted by the kubelet. This is hard to be replicated by exposing more
+information to the scheduler which may be too costly and subject to race
+conditions.
+
+### Goals
+
+- Allow deployers of fully managed worker nodes to have control over Pods
+  running on those nodes.
+
+- Enable node-local Pod admission checks without requiring changes to kubelet.
+
+### Non-Goals
+
+- Move existing validations to “out of tree” plugins.
+
+- Change existing API-layer validation solutions such as Pod Security Policies
+  and validating admission webhooks.
+
+## Proposal
+
+The approach taken is similar to the container networking interface (CNI) plugin
+architecture. With CNI, kubelet invokes one or more CNI plugin binaries on the
+host to set up a Pod’s networking. kubelet discovers available CNI plugins
+by [examining](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L242)
+a well-known directory (`/etc/cni/net.d`) for configuration files
+and [loading](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L248)
+plugin [descriptors](https://github.com/kubernetes/kubernetes/blob/f4db8212be53c69a27d893d6a4111422fbce8008/pkg/kubelet/dockershim/network/plugins.go#L52)
+upon startup.
+
+To support pluggable validation for pod admission on the worker node, we propose
+to have kubelet similarly discover node-local Pod admission plugins listed in a
+new PodAdmissionPluginDir flag.
+
+Other option is to enable this functionality through feature flag
+“enablePodAdmissionPlugin” and have the directory path defined inside the
+kubelet itself.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+These out-of-tree plugins should be provisioned and managed along with kubelet.
+
+## Design Details
+
+### Design
+
+#### Configuration file
+
+Node-local Pod admission plugins will be listed in a configuration file. The
+Plugins field indicates the list of plugins to invoke before admitting the Pod.
+
+```
+{
+  "name": "admission-plugin",
+  "version": "0.1",
+  "plugins": [
+    {
+      "name": "sysctlcheck",
+      "type": "shell"
+    },
+    {
+      "name": "fargatecheck",
+      "type": "shell"
+    }
+  ]
+}
+```
+
+A node-local Pod admission plugin has the following structure:
+
+```
+
+// PluginType indicates type of the admission plugin
+type PluginType string
+
+const (
+    PluginTypeShell  PluginType = "shell"  // binary to execute.
```

Comment:
It's not clear why we need 3 types of plugins.  Why is gRPC not sufficient?

I would probably expect the configuration for this to look more like kubelet configuration than "on disk configuration".  It's possible that the different types of kubelet configuration that exist have complicated the perception of this, but if this config is part of the kubelet, the real questions are:

1. Is this a static config tied to the kubelet process lifecycle?  Why or why not?
2. What flexibility does disk configuration add?  Why is it necessary?
3. What flexibility does having 3 types of plugin add?  Why is it necessary?

Also, another question I have is:

1. Was it easy for you to understand what the current guidance for plugins to Kube components or the Kubelet is, or were you having to look at a bunch of different examples and guess at what the recommendation was?

If it's the latter, that seems like something sig-node needs to clarify in a KEP or other doc "recommended design and tradeoffs for kubelet extension" so that we can save people time inferring it.